src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Unix.cs

```diff
@@ -79,41 +81,116 @@ internal static ValueTask<int> ReadAtOffsetAsync(SafeFileHandle handle, Memory<b
+                    fileOffset += bytesWritten;
+                }
             }
         }
 
-        internal static unsafe long WriteGatherAtOffset(SafeFileHandle handle, IReadOnlyList<ReadOnlyMemory<byte>> buffers, long fileOffset)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int GetNumberOfBytesToWrite(int byteCount)
         {
-            MemoryHandle[] handles = new MemoryHandle[buffers.Count];
-            Span<Interop.Sys.IOVector> vectors = buffers.Count <= IovStackThreshold ? stackalloc Interop.Sys.IOVector[IovStackThreshold] : new Interop.Sys.IOVector[buffers.Count ];
+#if DEBUG
+            // In debug only, to assist with testing, simulate writing fewer than the requested number of bytes.
+            if (byteCount > 1 &&  // ensure we don't turn the read into a zero-byte read
+                byteCount < 512)  // avoid on larger buffers that might have a length used to meet an alignment requirement
+            {
+                byteCount /= 2;
+            }
+#endif
+            return byteCount;
+        }
 
-            long result;
-            try
+        internal static unsafe void WriteGatherAtOffset(SafeFileHandle handle, IReadOnlyList<ReadOnlyMemory<byte>> buffers, long fileOffset)
+        {
+            int buffersCount = buffers.Count;
+            if (buffersCount == 0)
             {
-                int buffersCount = buffers.Count;
-                for (int i = 0; i < buffersCount; i++)
-                {
-                    ReadOnlyMemory<byte> buffer = buffers[i];
-                    MemoryHandle memoryHandle = buffer.Pin();
-                    vectors[i] = new Interop.Sys.IOVector { Base = (byte*)memoryHandle.Pointer, Count = (UIntPtr)buffer.Length };
-                    handles[i] = memoryHandle;
-                }
+                return;
+            }
 
-                fixed (Interop.Sys.IOVector* pinnedVectors = &MemoryMarshal.GetReference(vectors))
+            var handles = new MemoryHandle[buffersCount];
+            Span<Interop.Sys.IOVector> vectors = buffersCount <= IovStackThreshold ?
+                stackalloc Interop.Sys.IOVector[IovStackThreshold] :
+                new Interop.Sys.IOVector[buffersCount];
+
+            try
+            {
+                int buffersOffset = 0, firstBufferOffset = 0;
+                while (true)
                 {
-                    result = Interop.Sys.PWriteV(handle, pinnedVectors, buffers.Count, fileOffset);
+                    long totalBytesToWrite = 0;
+
+                    for (int i = buffersOffset; i < buffersCount; i++)
+                    {
+                        ReadOnlyMemory<byte> buffer = buffers[i];
+                        totalBytesToWrite += buffer.Length;
+
+                        MemoryHandle memoryHandle = buffer.Pin();
+                        vectors[i] = new Interop.Sys.IOVector { Base = firstBufferOffset + (byte*)memoryHandle.Pointer, Count = (UIntPtr)buffer.Length };
```

Comment:
It seems that when `firstBufferOffset > 0` the `Count` might be invalid?

```diff
- vectors[i] = new Interop.Sys.IOVector { Base = firstBufferOffset + (byte*)memoryHandle.Pointer, Count = (UIntPtr)buffer.Length };
+ vectors[i] = new Interop.Sys.IOVector { Base = firstBufferOffset + (byte*)memoryHandle.Pointer, Count = (UIntPtr)(buffer.Length - firstBufferOffset) };
```

Reply:
Welcome back!

And, yes, good catch.
https://github.com/dotnet/runtime/pull/55909